core/pkg/ingress/types.go

```diff
@@ -0,0 +1,288 @@
+	// certificate and key.
+	DefaultSSLDirectory = "/ingress-controller/ssl"
+)
+
+// Controller holds the methods to handle an Ingress backend
+// TODO (#18): Make sure this is sufficiently supportive of other backends.
+type Controller interface {
+	// Reload takes a byte array representing the new loadbalancer configuration,
+	// and returns a byte array containing any output/errors from the backend.
+	// Before returning the backend must load the configuration in the given array
+	// into the loadbalancer and restart it, or fail with an error and message string.
+	// If reloading fails, there should be not change in the running configuration or
+	// the given byte array.
+	Reload(data []byte) ([]byte, error)
+	// Tests returns the commands to execute that verifies if the configuration file is valid
+	// Example: nginx -t -c <file>
+	Test(file string) *exec.Cmd
+	// OnUpdate callback invoked from the sync queue https://k8s.io/ingress/core/blob/master/pkg/ingress/controller/controller.go#L387
+	// when an update occurs. This is executed frequently because Ingress
+	// controllers watches changes in:
+	// - Ingresses: main work
+	// - Secrets: referenced from Ingress rules with TLS configured
+	// - ConfigMaps: where the controller reads custom configuration
+	// - Services: referenced from Ingress rules and required to obtain
+	//	 information about ports and annotations
+	// - Endpoints: referenced from Services and what the backend uses
+	//	 to route traffic
+	// Any update to services, endpoints, secrets (only those referenced from Ingress)
+	// and ingress trigger the execution.
+	// Notifications of type Add, Update and Delete:
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/client/cache/controller.go#L164
+	//
+	// ConfigMap content of --configmap
+	// Configuration returns the translation from Ingress rules containing
+	// information about all the upstreams (service endpoints ) "virtual"
+	// servers (FQDN) and all the locations inside each server. Each
+	// location contains information about all the annotations were configured
+	// https://k8s.io/ingress/core/blob/master/pkg/ingress/types.go#L83
+	// The backend returns the contents of the configuration file or an error
+	// with the reason why was not possible to generate the file.
+	//
+	// The returned configuration is then passed to test, and then to reload
+	// if there is no errors.
+	OnUpdate(*api.ConfigMap, Configuration) ([]byte, error)
+	// BackendDefaults returns the minimum settings required to configure the
+	// communication to endpoints
+	BackendDefaults() defaults.Backend
+	// Info returns information about the ingress controller
+	Info() *BackendInfo
+}
+
+// BackendInfo returns information about the backend.
+// This fields contains information that helps to track issues or to
+// map the running ingress controller to source code
+type BackendInfo struct {
+	// Name returns the name of the backend
```

Comment:
This comment could use some clarification since it's required not optional, what does name imply? does it follow some format, or can it be a random hash/alphanumerics etc